docs/serving/feature-flags.md

```diff
@@ -0,0 +1,68 @@
+---
+title: "Feature/Extension Flags"
+weight: 50
+type: "docs"
+---
+
+Knative is deliberate about the concepts it incorporates into its API. The API aims to be portable and abstracts away the specificities of an implementation. That being said, the Knative users should be empowered to use the useful features provided by their platform of choice implementing the Knative API.
+
+This document introduces two concepts:
+* Feature: a way to stage the introduction of features to the Knative API.
+* Extension: a way to extend Knative beyond the portable concepts of the Knative API.
+
+## Control
+Features and extensions are controlled by flags defined in the `config-features` ConfigMap in the `knative-serving` namespace.
+Flags can have the following values:
+* Enabled: the feature is enabled.
+* Allowed: the feature can be enabled (e.g. using an annotation).
+* Disabled: the feature cannot be enabled.
+
+These three states don't make sense for all features.
+Let's consider two types of features: `multi-container` and `kubernetes/podspec-dryrun`.
+
+`multi-container` allows the user to specify more than one container in the Knative Service spec. In this case, `Enabled` and `Allowed` are equivalent because using this feature requires to actually use it in the Knative Service spec. If a single container is specified, whether the feature is enabled or not doesn't change anything.
+
+`kubernetes/podspec-dryrun` changes the behavior of the Kubernetes implementation of the Knative API, but it has nothing to do with the Knative API itself. In this case, `Enabled` means the feature will be enabled unconditionally, `Allowed` means that the feature will be enabled only when specified with an annotation and `Disabled` means that the feature cannot be used at all.
+
+## Lifecyle
+Features and extensions go through 3 similar phases (Alpha, Beta, GA) but with important differences.
+
+Alpha means:
+* Might be buggy. Enabling the feature may expose bugs.
+* Support for feature may be dropped at any time without notice.
+* The API may change in incompatible ways in a later software release without notice.
+* Recommended for use only in short-lived testing clusters, due to increased risk of bugs and lack of long-term support.
+
+Beta means:
+* The feature is well tested. Enabling the feature is considered safe.
+* Support for the overall feature will not be dropped, though details may change.
```

Comment:
But then there is no real difference between Alpha & Beta.
BTW, that's the k8s model https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/

Reply:
Also, it forces user to explicitly enable it.
For example, a feature where Enabled and Allowed have the same meaning, like `multi-container`:
* if it is `Disabled` in Alpha, a user has to enable it manually
* if it is `Allowed` in Alpha, a user could use the feature by accident without realizing it (by simply adding another container to their Ksvc spec).

It is safer this way.

Reply:
Oh I just mean that it's more like:

Alpha: The feature may be dropped without notice
Beta: The feature is committed to the same deprecation policy as fully released features.

The language "the overall feature will not be dropped" makes it sound like it can never ever be deprecated and all features live forever.